docs/src/main/tut/kleisli.md

```diff
@@ -0,0 +1,237 @@
+
+val reciprocal = Kleisli((i: Int) => if (i == 0) None else Some(1.0 / i))
+
+val parseAndReciprocal = reciprocal.compose(parse)
+```
+
+`Kleisli#andThen` can be defined similarly.
+
+It is important to note that the `F[_]` having a `FlatMap` (or a `Monad`) instance is not a hard requirement -
+we can do useful things with weaker requirements. Such an example would be `Kleisli#map`, which only requires
+that `F[_]` have a `Functor` instance (e.g. is equipped with `map: F[A] => (A => B) => F[B]`).
+
+```tut
+import cats.Functor
+
+final case class Kleisli[F[_], A, B](run: A => F[B]) {
+  def map[C](f: B => C)(implicit F: Functor[F]): Kleisli[F, A, C] =
+    Kleisli[F, A, C](a => F.map(run(a))(f))
+}
+```
+
+### Type class instances
+The type class instances for `Kleisli`, like that for functions, fix the input type (and the `F[_]`) and leave
+the output type free. What type class instances it has tends to depend on what instances the `F[_]` has. For
+instance, `Kleisli[F, A, B]` has a `Functor` instance so long as the chosen `F[_]` does. It has a `Monad`
+instance so long as the chosen `F[_]` does. The instances in Cats are laid out in a way such that implicit
+resolution will pick up the most specific instance it can (depending on the `F[_]`).
+
+An example of a `Monad` instance for `Kleisli` would be:
+
+```tut
+// We can define a FlatMap instance for Kleisli if the F[_] we chose has a FlatMap instance
+// Note the input type and F are fixed, with the output type left free
+implicit def kleisliFlatMap[F[_], Z](implicit F: FlatMap[F]): FlatMap[Kleisli[F, Z, ?]] =
+  new FlatMap[Kleisli[F, Z, ?]] {
+    def flatMap[A, B](fa: Kleisli[F, Z, A])(f: A => Kleisli[F, Z, B]): Kleisli[F, Z, B] =
+      Kleisli(z => F.flatMap(fa.run(z))(a => f(a).run(z)))
+
+    def map[A, B](fa: Kleisli[F, Z, A])(f: A => B): Kleisli[F, Z, B] =
+      Kleisli(z => F.map(fa.run(z))(f))
+  }
+```
+
+## Other uses
+### Monad Transformer
+Many data types have a monad transformer equivalent that allows us to compose the `Monad` instance of the data
+type with any other `Monad` instance. For instance, `OptionT[F[_], A]` allows us to compose the monadic properties
+of `Option` with any other `F[_]`, such as a `List`. This allows us to work with nested contexts/effects in a
+nice way, say, in for comprehensions.
+
+`Kleisli` can be viewed as the monad transformer for functions. Recall that at its essence, `Kleisli[F, A, B]`
+is just a function `A => F[B]`, with niceties to make working with the value we actually care about, the `B`, easy.
+`Kleisli` allows us to take the effects of functions and have them play nice with the effects of any other `F[_]`.
+
+This may raise the question, what exactly is the "effect" of a function?
+
+Well, if we take a look at any function, we can see it takes some input and produces some output with it, without
+having touched the input (assuming a functional environment). That is, we take a read-only value, and produce some
```

Comment:
Maybe instead of "assuming a functional environment" we could say "assuming the function is [referentially transparent](https://en.wikipedia.org/wiki/Referential_transparency_%28computer_science%29)" ?